### PR TITLE
Deprecate overriding DiscoveryPlugin internals

### PR DESCRIPTION
**Pull request by rjernst**
DiscoveryPlugin allows extending getJoinValidator and
getElectionStrategies. These are implementation details of the system.
This commit deprecates these methods so that plugin authors
are discouraged from overriding them.

### CODE REVIEW

File: docs/reference/migration/migrate_8_5.asciidoc [hunk 1]
```diff
@@ -20,3 +20,38 @@ coming::[8.5.0]
...
 // end::notable-breaking-changes[]
 
+[discrete]
+[[deprecated-8.5]]
+=== Deprecations
+
+The following functionality has been deprecated in {es} 8.5
+and will be removed in a future version.
+While this won't have an immediate impact on your applications,
+we strongly encourage you take the described steps to update your code
```
Comment (grcevski) [inline docs/reference/migration/migrate_8_5.asciidoc:30]: missing 'to' between you and take.

**Review by grcevski — APPROVED** [review]
LGTM! One minor typo and it seems the YML area needs to be fixed. We possibly want to add an area for the plugins as we'll do bunch of this?

**Comment by rjernst** [comment]
> We possibly want to add an area for the plugins as we'll do bunch of this?

Yes, @tvernum is going to followup with the docs team to add that. In the mean time, I will use `Java API` here.

**Review by qhoxie — APPROVED** [review]

**Comment by rjernst** [comment]
@elasticsearchmachine run elasticsearch-ci/packaging-tests-unix-sample

**Comment by rjernst** [comment]
@elasticsearchmachine run elasticsearch-ci/packaging-tests-unix-sample

**Pull request merged**